src/hotspot/share/gc/shenandoah/shenandoahMarkingContext.cpp

```diff
@@ -25,63 +26,71 @@
 #include "precompiled.hpp"
 #include "gc/shared/markBitMap.inline.hpp"
 #include "gc/shenandoah/shenandoahHeap.inline.hpp"
-#include "gc/shenandoah/shenandoahHeapRegion.inline.hpp"
 #include "gc/shenandoah/shenandoahMarkingContext.hpp"
-#include "gc/shenandoah/shenandoahTaskqueue.inline.hpp"
-#include "utilities/stack.inline.hpp"
 
-ShenandoahMarkingContext::ShenandoahMarkingContext(MemRegion heap_region, MemRegion bitmap_region, size_t num_regions, uint max_queues) :
+ShenandoahMarkingContext::ShenandoahMarkingContext(MemRegion heap_region, MemRegion bitmap_region, size_t num_regions) :
   _mark_bit_map(heap_region, bitmap_region),
   _top_bitmaps(NEW_C_HEAP_ARRAY(HeapWord*, num_regions, mtGC)),
   _top_at_mark_starts_base(NEW_C_HEAP_ARRAY(HeapWord*, num_regions, mtGC)),
   _top_at_mark_starts(_top_at_mark_starts_base -
-                      ((uintx) heap_region.start() >> ShenandoahHeapRegion::region_size_bytes_shift())),
-  _task_queues(new ShenandoahObjToScanQueueSet(max_queues)) {
-  assert(max_queues > 0, "At least one queue");
-  for (uint i = 0; i < max_queues; ++i) {
-    ShenandoahObjToScanQueue* task_queue = new ShenandoahObjToScanQueue();
-    _task_queues->register_queue(i, task_queue);
-  }
-}
-
-ShenandoahMarkingContext::~ShenandoahMarkingContext() {
-  for (uint i = 0; i < _task_queues->size(); ++i) {
-    ShenandoahObjToScanQueue* q = _task_queues->queue(i);
-    delete q;
-  }
-  delete _task_queues;
+                      ((uintx) heap_region.start() >> ShenandoahHeapRegion::region_size_bytes_shift())) {
 }
 
 bool ShenandoahMarkingContext::is_bitmap_clear() const {
   ShenandoahHeap* heap = ShenandoahHeap::heap();
   size_t num_regions = heap->num_regions();
   for (size_t idx = 0; idx < num_regions; idx++) {
     ShenandoahHeapRegion* r = heap->get_region(idx);
-    if (heap->is_bitmap_slice_committed(r) && !is_bitmap_clear_range(r->bottom(), r->end())) {
+    if (r->is_affiliated() && heap->is_bitmap_slice_committed(r) && !is_bitmap_clear_range(r->bottom(), r->end())) {
       return false;
     }
   }
   return true;
 }
 
-bool ShenandoahMarkingContext::is_bitmap_clear_range(HeapWord* start, HeapWord* end) const {
-  return _mark_bit_map.get_next_marked_addr(start, end) == end;
+bool ShenandoahMarkingContext::is_bitmap_clear_range(const HeapWord* start, const HeapWord* end) const {
+  if (start < end) {
+    ShenandoahHeap* heap = ShenandoahHeap::heap();
+    size_t start_idx = heap->heap_region_index_containing(start);
+    size_t end_idx = heap->heap_region_index_containing(end - 1);
+    while (start_idx <= end_idx) {
+      ShenandoahHeapRegion* r = heap->get_region(start_idx);
+      if (!heap->is_bitmap_slice_committed(r)) {
+        return true;
+      }
+      start_idx++;
+    }
+  }
+  return _mark_bit_map.is_bitmap_clear_range(start, end);
```

Comment:
Comprehension check: this seems to bail out the moment it discovers an uncommitted slice. Does it really happen? More worryingly, if there is a mix of committed and uncommitted chunks, this method returns `true`, even if committed chunks are not actually clean?

Reply:
In all (3) usages of this method, the `start` and `end` are within the same region, so it isn't really iterating across regions. I'll rewrite it to handle this case more clearly.

Reply:
https://bugs.openjdk.org/browse/JDK-8344797